### ISSUE TITLE
CGAL_assertion(false)/CGAL_error() replace by CGAL_unreachable()?

### ISSUE THREAD

**Issue by lrineau** [CGAL/cgal#8295]
There are more than 250 occurrences of `CGAL_assertion(false)` in CGAL. And other occurrences of `CGAL_error()`. Should not we replace them with `CGAL_unreachable()`?

**Comment by afabri** [comment]
Is there really something to discuss?  I guess it must be dealt with case by case and cannot be a global replace.

**Comment by afabri** [comment]
After an assertion we in general still return something in case the function has a return value.  Is this also the case when replacing with the unreachable statement?

**Comment by afabri** [comment]
Can a `default:` in a `switch` be at the very beginning (see [here](https://github.com/CGAL/cgal/blob/7570ad8a43994b2890f0172293f48e69ca803672/Periodic_2_triangulation_2/include/CGAL/Periodic_2_Delaunay_triangulation_2.h#L475))? Does it then still go to the other cases?

**Comment by afabri** [comment]
Do you agree that we make [this](https://github.com/CGAL/cgal/blob/7570ad8a43994b2890f0172293f48e69ca803672/Periodic_3_mesh_3/include/CGAL/Periodic_3_mesh_3/Protect_edges_sizing_field.h#L1099) to a `CGAL_assertion_msg()`?  
With assertions switched off, we then no longer see something on `cerr`, but that is not desired anyways.

**Comment by afabri** [comment]
Do you agree that [here](https://github.com/CGAL/cgal/blob/7570ad8a43994b2890f0172293f48e69ca803672/Surface_mesh_deformation/include/CGAL/Surface_mesh_deformation.h#L568) we cannot replace because it is something that can happen ?

**Comment by sloriot** [comment]
> Do you agree that [here](https://github.com/CGAL/cgal/blob/7570ad8a43994b2890f0172293f48e69ca803672/Surface_mesh_deformation/include/CGAL/Surface_mesh_deformation.h#L568) we cannot replace because it is something that can happen ?

assertion should be removed

**Comment by afabri** [comment]
Does it make sense to have an assertion in [`TDS_3::is_valid()`](https://github.com/CGAL/cgal/blob/7570ad8a43994b2890f0172293f48e69ca803672/TDS_3/include/CGAL/Triangulation_data_structure_3.h#L3595)?

**Comment by afabri** [comment]
Shall we make [here](https://github.com/CGAL/cgal/blob/7570ad8a43994b2890f0172293f48e69ca803672/Tetrahedral_remeshing/include/CGAL/Tetrahedral_remeshing/internal/tetrahedral_adaptive_remeshing_impl.h#L503) the last `case`  to `default` and then we don't need an assert or unreachable statement.

**Comment by sloriot** [comment]
> Does it make sense to have an assertion in [`TDS_3::is_valid()`](https://github.com/CGAL/cgal/blob/7570ad8a43994b2890f0172293f48e69ca803672/TDS_3/include/CGAL/Triangulation_data_structure_3.h#L3595)?

I would remove it

**Comment by afabri** [comment]
> > Do you agree that [here](https://github.com/CGAL/cgal/blob/7570ad8a43994b2890f0172293f48e69ca803672/Surface_mesh_deformation/include/CGAL/Surface_mesh_deformation.h#L568) we cannot replace because it is something that can happen ?
> 
> assertion should be removed

So _unreachable_ means that one may still arrive there in practice?

**Comment by sloriot** [comment]
> > > Do you agree that [here](https://github.com/CGAL/cgal/blob/7570ad8a43994b2890f0172293f48e69ca803672/Surface_mesh_deformation/include/CGAL/Surface_mesh_deformation.h#L568) we cannot replace because it is something that can happen ?
> > 
> > 
> > assertion should be removed
> 
> So _unreachable_ means that one may still arrive there in practice?

In this case, it emits an assertion while returning `false` is the expected behavior.

**Comment by afabri** [comment]
> > > > Do you agree that [here](https://github.com/CGAL/cgal/blob/7570ad8a43994b2890f0172293f48e69ca803672/Surface_mesh_deformation/include/CGAL/Surface_mesh_deformation.h#L568) we cannot replace because it is something that can happen ?
> > > 
> > > 
> > > assertion should be removed
> > 
> > 
> > So _unreachable_ means that one may still arrive there in practice?
> 
> In this case, it emits an assertion while returning `false` is the expected behavior.

But when we may arrive at that point, we maybe want to get an exception thrown by a custom error handler.

**Comment by afabri** [comment]
Shall we declare [this](https://github.com/CGAL/cgal/blob/418b00e27469c6ed8b2539242e5158d5ece072ca/Mesh_3/include/CGAL/Mesh_3/C3T3_helpers.h#L76) copy constructor private instead?

**Comment by sloriot** [comment]
> > > > > Do you agree that [here](https://github.com/CGAL/cgal/blob/7570ad8a43994b2890f0172293f48e69ca803672/Surface_mesh_deformation/include/CGAL/Surface_mesh_deformation.h#L568) we cannot replace because it is something that can happen ?
> > > > 
> > > > 
> > > > assertion should be removed
> > > 
> > > 
> > > So _unreachable_ means that one may still arrive there in practice?
> > 
> > 
> > In this case, it emits an assertion while returning `false` is the expected behavior.
> 
> But when we may arrive at that point, we maybe want to get an exception thrown by a custom error handler.

you can assume it!=end since there is a check before

**Comment by afabri** [comment]
Would this also compile if the static function [here ](https://github.com/CGAL/cgal/blob/418b00e27469c6ed8b2539242e5158d5ece072ca/Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Functors_for_face_graph_wrapper.h#L42) was just removed?   After all we need the class only because of the offered specializations.

**Comment by sloriot** [comment]
> Would this also compile if the static function [here ](https://github.com/CGAL/cgal/blob/418b00e27469c6ed8b2539242e5158d5ece072ca/Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Functors_for_face_graph_wrapper.h#L42) was just removed? After all we need the class only because of the offered specializations.

Indeed, you should get a compilation error instead of a runtime error.

**Comment by afabri** [comment]
> > Would this also compile if the static function [here ](https://github.com/CGAL/cgal/blob/418b00e27469c6ed8b2539242e5158d5ece072ca/Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Functors_for_face_graph_wrapper.h#L42) was just removed? After all we need the class only because of the offered specializations.
> 
> Indeed, you should get a compilation error instead of a runtime error.

To make it very clear we could even declare the constructor `=0`.

**Comment by lrineau** [comment]
> Shall we declare [this](https://github.com/CGAL/cgal/blob/418b00e27469c6ed8b2539242e5158d5ece072ca/Mesh_3/include/CGAL/Mesh_3/C3T3_helpers.h#L76) copy constructor private instead?

Make it `= delete;` instead.

**Comment by lrineau** [comment]
> Would this also compile if the static function [here ](https://github.com/CGAL/cgal/blob/418b00e27469c6ed8b2539242e5158d5ece072ca/Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Functors_for_face_graph_wrapper.h#L42) was just removed? After all we need the class only because of the offered specializations.

I agree it should be remove. Once it is remove, if a user-code uses `Get_beta<Face_graph, i>` for `i>3`, there would be a compilation error.

**Comment by lrineau** [comment]
There should be only a declaration of the template, without any definition. Just:

```c++
template<typename HEG, unsigned int i> struct Get_beta;
```

**Comment by afabri** [comment]
[Here](https://github.com/CGAL/cgal/blob/50219fc33bc53b5846e9a4257befc6baef3f395a/Convex_decomposition_3/include/CGAL/Convex_decomposition_3/Single_wall_creator.h#L257) we have the assertion as first line of a function.

**Comment by afabri** [comment]
Is there a sublelty I miss here which makes that there is a `CGAL_assertion(false)` at the end, while before we have several `CGAL_unreachable()` @sloriot

**Comment by sloriot** [comment]
> [Here](https://github.com/CGAL/cgal/blob/50219fc33bc53b5846e9a4257befc6baef3f395a/Convex_decomposition_3/include/CGAL/Convex_decomposition_3/Single_wall_creator.h#L257) we have the assertion as first line of a function.

git blame says that the code is 16yo and from Peter. So basically it was always there. We can remove the function.

**Comment by afabri** [comment]
> > [Here](https://github.com/CGAL/cgal/blob/50219fc33bc53b5846e9a4257befc6baef3f395a/Convex_decomposition_3/include/CGAL/Convex_decomposition_3/Single_wall_creator.h#L257) we have the assertion as first line of a function.
> 
> git blame says that the code is 16yo and from Peter. So basically it was always there. We can remove the function.

Well, it is called [here](https://github.com/CGAL/cgal/blob/50219fc33bc53b5846e9a4257befc6baef3f395a/Convex_decomposition_3/include/CGAL/Convex_decomposition_3/Single_wall_creator.h#L355).

**Comment by afabri** [comment]
@sloriot any idea what the function [`initialize_index_map()`](https://github.com/CGAL/cgal/blob/119bd102c9f8d77dc80d09f827381c504dba1490/BGL/include/CGAL/boost/graph/internal/initialized_index_maps_helpers.h#L84)   is about?   I do not see it used anywhere. Do we have it for backward compatibility?

**Comment by sloriot** [comment]
> @sloriot any idea what the function [`initialize_index_map()`](https://github.com/CGAL/cgal/blob/119bd102c9f8d77dc80d09f827381c504dba1490/BGL/include/CGAL/boost/graph/internal/initialized_index_maps_helpers.h#L84) is about? I do not see it used anywhere. Do we have it for backward compatibility?

probably a leftover after some refactoring.

**Comment by afabri** [comment]
@MaelRL does "meaningless" [here](https://github.com/CGAL/cgal/blob/119bd102c9f8d77dc80d09f827381c504dba1490/Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/locate.h#L1193) that this cannot happen for whatever input, or does it mean that the user called `locate()` with illegal input?

**Comment by MaelRL** [comment]
> @MaelRL does "meaningless" [here](https://github.com/CGAL/cgal/blob/119bd102c9f8d77dc80d09f827381c504dba1490/Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/locate.h#L1193) that this cannot happen for whatever input, or does it mean that the user called `locate()` with illegal input?

illegal input, this is a violation of the precondition:
```
/// \pre `loc` corresponds to a point that lies on a face incident to both `loc.first` and `fd`.
```

**Comment by MaelRL** [comment]
> @MaelRL [here](https://github.com/CGAL/cgal/blob/119bd102c9f8d77dc80d09f827381c504dba1490/Periodic_3_triangulation_3/include/CGAL/Periodic_3_triangulation_3/internal/Periodic_3_triangulation_iterators_3.h#L89) I would prefer to see added the `case T::NONE` which is the default. Can it by construction not be `NONE`? Then I will put `CGAL_unreachable()`

OK

**Comment by afabri** [comment]
@efifogel can you help me to understand the design idea for the generic operator [here](https://github.com/CGAL/cgal/blob/119bd102c9f8d77dc80d09f827381c504dba1490/Arrangement_on_surface_2/include/CGAL/Arr_point_location/Trapezoidal_decomposition_2.h#L508), which is supposed not to be used.

**Comment by afabri** [comment]
@gdamiand can you help me to understand the design idea for the function `run()`   [here](https://github.com/CGAL/cgal/blob/119bd102c9f8d77dc80d09f827381c504dba1490/Combinatorial_map/include/CGAL/Combinatorial_map/internal/Combinatorial_map_group_functors.h#L894) , which is supposed not to be used.

**Comment by gdamiand** [comment]
> @gdamiand can you help me to understand the design idea for the function `run()` [here](https://github.com/CGAL/cgal/blob/119bd102c9f8d77dc80d09f827381c504dba1490/Combinatorial_map/include/CGAL/Combinatorial_map/internal/Combinatorial_map_group_functors.h#L894) , which is supposed not to be used.

The functor Test_split_nonvoid_attribute_functor_run<CMap, i, j> tests if i-attributes are split after an operation, except for j-attributes.

As said in the comments, for j==0 or j==1, we use only the version with two lists of darts, thus this run method is never called (but should exist for the compiler).
